[빅스] 코틀린 자동차 경주 제출합니다.

briandr97 opened the pull request:
미션 완료하고 리뷰 요청드립니다!

아래 목록은 이번 미션을 수행하면서 특히 궁금했던 부분들입니다..!

- object를 사용했는데 언제 object를 사용하는 것이 좋을지 궁금합니다!
- GameController는 input과 output 모두 수행되는데 어떻게 test 해야 할지 (테스트가 필요한 부분일지) 궁금합니다!
- GameController에 모든 것이 합쳐지면서 비대해지며, 순수 함수로 작성하기가 어려워 테스트가 어려워지는 것 같습니다. 현재 GameController가 적절한 양의 책임을 갖고 있는 것인지 궁금합니다.
- 역할을 잘 분리한 것인지 궁금합니다!
- 테스트하는 방향성이 맞는지 궁금합니다!

감사합니다!

briandr97 committed:
docs(README): 구현할 기능 목록 작성

briandr97 committed:
feat(domain): 도메인 패키지 기초 세팅

briandr97 committed:
feat(view): view 패키지 기초 세팅

briandr97 committed:
feat(InputView): Create basic method in InputView

briandr97 committed:
feat(OutputView): Create basic method in OutputView

briandr97 committed:
docs(README): 구현한 기능 목록 수정

briandr97 committed:
feat(NumberGenerator, RandomNumberGenerator): 랜덤 번호 생성기, 구현체 추가

briandr97 committed:
feat(OutputView): 함수, 변수명 변경

briandr97 committed:
feat(Driver): 운전자(전진, 정지 판독기) 추가

briandr97 committed:
feat(DriverDecision): 전진, 정지에 대한 Enum 클래스 추가

briandr97 committed:
feat(InputView): readLine으로 null 값이 들어왔을 때에 대한 에러 처리

briandr97 committed:
feat(DriverDecision): 전진, 정지에 대한 enum 클래스에 distance 프로퍼티 추가

briandr97 committed:
feat(Referee): 우승자를 판단해줄 Referee 클래스 구현

briandr97 committed:
feat(Driver): 함수명 및 구조 변경

 - numberGenerator 주입 받기
 - decideMovement로 함수명 변경

briandr97 committed:
feat(Car): 이름과 이동거리를 갖고 있는 Car 클래스 구현

briandr97 committed:
feat(OutputView): printResult의 매개변수 자료형 변경

briandr97 committed:
docs(README): 수행한 목록 업데이트

briandr97 committed:
feat(Application): Create Application kotlin file

1. It is game entry point

briandr97 committed:
feat(Car): Create some blocks

1. Create Except method in init block
2. Create companion object for hard coding

briandr97 committed:
feat(GameController): Create gameContorller Class

1. It has methods of game flow

briandr97 committed:
refactor(InputView): Modify enterTryCount exception control

briandr97 committed:
feat(OutputView): Create printErrorMessage method

briandr97 committed:
feat(CarTest): Create CarTest Class

1. Test car movement
2. Test car name validation

briandr97 committed:
feat(TestCarFactory): Make car for test

briandr97 committed:
feat(RefereeTest): Create RefereeTest class

1. Test judge winners

briandr97 committed:
feat(DriverTest): Create DriverTest class

1. Test decideMovement method

briandr97 committed:
feat(TestNumberGenerator): Create TestNumberGenerator for Test Class

briandr97 committed:
Refactor(CarTest): Create CarTest class

1. Change CarFactory name

briandr97 committed:
Refactor(Car): Release private of value NAME_CONVENTION_ERROR_MESSAGE

briandr97 committed:
docs(README): 수행한 한 일 목록 수정

briandr97 committed:
feat(DriverTest): change class to internal class

galcyurio commented on src/main/kotlin/domain/Referee.kt:
```diff
@@ -0,0 +1,23 @@
+package domain
+
+object Referee {
```
> object를 사용했는데 언제 object를 사용하는 것이 좋을지 궁금합니다!

상태를 가지지 않는 경우에 사용됩니다.
개인적으로는 보통 Util성 함수들을 선언하는 경우에만 사용하는 편이에요.

현재는 상태가 없더라도 조금씩 유지보수를 하다보면 상태를 가지게 되거나 생성자로 주입해야하는 의존성이 생기거나 List로 관리해야하는 경우가 생겨버리기 때문이에요.

이건 저만의 기준이니 object를 언제 활용하는지는 여러 곳을 리서치해보시는걸 추천드립니다.

galcyurio commented on src/main/kotlin/GameController.kt:
```diff
@@ -0,0 +1,75 @@
+
+private const val EXCEPTION_NULL = "[ERROR] exception 메시지가 null입니다."
+
+class GameController(
```
>GameController는 input과 output 모두 수행되는데 어떻게 test 해야 할지 (테스트가 필요한 부분일지) 궁금합니다!

> GameController에 모든 것이 합쳐지면서 비대해지며, 순수 함수로 작성하기가 어려워 테스트가 어려워지는 것 같습니다. 현재 GameController가 적절한 양의 책임을 갖고 있는 것인지 궁금합니다.

이번 미션에서 View는 테스트하지 않으셔도 됩니다.

보통 반환형이 없는 함수가 테스트가 어려운데 이런 함수들이 왜 테스트가 어려운 것인지 어떻게 개선할 수 있을지에 대해서는 고민해보시면 좋겠습니다.
매번 출력하는게 아니라 녹화한다는 개념으로 무언가를 반환해보는건 어떨까요?

print 때문에 순수함수로 바꿀 수 없어서 테스트하기 힘든 코드들이 많다면 print를 최소화하는 방법을 생각해보세요.

galcyurio commented on README.md:
```diff
@@ -1 +1,17 @@
 # 자동차 경주 게임
```
> 역할을 잘 분리한 것인지 궁금합니다!

네, 과도하지 않게 잘 분리해주셨습니다.

galcyurio commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -0,0 +1,51 @@
+
+    @MethodSource("provideCar")
+    @ParameterizedTest
+    fun `차 주행 테스트`(randNum: Int, expectedDistance: String) {
```
> 테스트하는 방향성이 맞는지 궁금합니다!

테스트 코드는 괜찮은데 전체적으로 테스트 케이스 이름이 명확하지 않습니다.
테스트 케이스 이름만 보아도 이 프로그램이 어떻게 동작하는지 알 수 있도록 만들어보세요.

galcyurio commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -0,0 +1,51 @@
+        when (carException == null) {
+            true -> testSuccessCase(car, carName)
+            false -> testFailCase(carException)
+        }
```
테스트 코드에서는 분기처리가 나오지 않아야 합니다.
만약 분기처리가 필요하다고 생각되면, 2개의 테스트 케이스가 되어야 하는게 아닌지 고민해보세요.

galcyurio commented on src/test/kotlin/util/TestCarFactory.kt:
```diff
@@ -0,0 +1,24 @@
+import domain.Car
+import domain.Driver
+
+object TestCarFactory {
+    fun makeCars(carNames: List<String>, randNums: List<Int>): List<Car> {
+        return carNames.mapIndexed { idx, name ->
+            Car(name, makeDriver(randNums[idx]))
+        }
+    }
+
+    fun makeCars(carNames: List<String>, randNum: Int): List<Car> {
+        return carNames.map { Car(it, makeDriver(randNum)) }
+    }
+
+    fun makeCar(carName: String, randNum: Int): Car {
+        return Car(carName, makeDriver(randNum))
+    }
+
+    private fun makeDriver(randNum: Int): Driver {
+        return Driver(TestNumberGenerator(randNum))
+    }
+}
```
가독성을 위해 크게 중요하지 않은 코드들을 잘 분리해주셨습니다 :+1:

galcyurio commented on src/main/kotlin/util/DriverDecision.kt:
```diff
@@ -0,0 +1,6 @@
+package util
```
차량이 움직일지를 결정하는 중요한 도메인이기 때문에 domain 패키지에 위치해야 합니다.

galcyurio commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,23 @@
+package domain
+
+class Car(val carName: String, private val driver: Driver) {
```
차량이 driver를 가지고 있네요. 설계가 현실같아서 재미있네요 :)

galcyurio commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,23 @@
+        require(carName.length in RANGE) { NAME_CONVENTION_ERROR_MESSAGE }
+    }
+
+    fun decideCarMovement() {
+        val decision = driver.decideMovement()
+        _distance += decision.distance
+    }
```
움직일지 판단하는 것은 Driver가 처리했고 차량 입장에서는 움직이는 것이니까 `move()` 와 같은 함수명이 더 어울리지 않을까요?

galcyurio commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    companion object {
+        const val NAME_CONVENTION_ERROR_MESSAGE = "[ERROR] 이름은 1~5글자여야합니다."
```
아래의 상수인 1, 5가 변경되는 경우 이 메시지도 같이 변경되도록 만들어보세요!

galcyurio reviewed (changes requested):
1단계 자동차 미션 잘 완료해주셨습니다
몇 가지 피드백을 남겨두었으니 확인 후 반영해주세요 :)

briandr97 committed:
refactor(Referee) filter와 map 함수를 이용하여 불필요한 코드를 줄임

briandr97 committed:
refactor(Car): backing field로 처리한 distance를 overrided setter로 처리

원래는 distance를 외부에 보이돼 수정은 내부에서만 가능하도록 backing field를 이용해 처리했었다.
그러나 List, MutableList처럼 mutable과 immutable로 자료형이 나뉘어진 경우가 아니라면 setter를 override했을 때 같은 효과를 보면서 변수도 하나 줄일 수 있기 때문에 더 좋다고 생각돼어 변경하였다.

briandr97 committed:
refactor(Car): 상수화한 에러메시지에 확장성 추가

1, 5를 하드코딩하지 않고 다른 상수로 넣었다.

briandr97 committed:
refactor(Car): 함수명 move로 변경

briandr97 committed:
refactor(domain): 패키지 구조 변경

- DriverDecision을 util에서 domain으로 패키지 변경
- 숫자 생성관련 파일들을 numbergenerator 패키지로 변경

briandr97 committed:
refactor(CarTest): 한 테스트의 분기처리를 두 개의 테스트로 분리, 함수명 변경

briandr97 committed:
refactor(GameController): 게임의 전체적인 진행을 GameController에서 관장하고 main함수는 그저 Controller를 실행시키기만 한다.

briandr97 committed:
refactor(view): InputView, OutputView를 object로 바꿈

galcyurio commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -11,34 +11,33 @@ internal class CarTest {
-    }
-
-    private fun testFailCase(carException: Throwable) {
+    fun `차 이름이 1~5 범위 밖인 경우`(carName: String) {
```
https://github.com/woowacourse/kotlin-racingcar/pull/54#discussion_r1101765095

테스트 케이스 이름들이 모두 명확하지 않습니다.
차 이름이 1~5 범위 안, 밖인 경우 어떻게 되어야 하는지를 테스트 함수명에 포함시켜주세요.
요구사항을 그대로 테스트 케이스 이름으로 만들어도 좋습니다.

galcyurio commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -11,34 +11,33 @@ internal class CarTest {
+    fun `차 이름이 1~5 범위 밖인 경우`(carName: String) {
+        // given
+        // when
+        val carException = kotlin.runCatching { TestCarFactory.makeCar(carName, 5) }.exceptionOrNull()
```
자동완성이 잘못 되었네요. `kotlin.`을 떼고 `runCatching()`으로 import 없이 호출할 수 있습니다.

galcyurio reviewed (changes requested):
피드백 잘 반영해주셨네요 🙂 
테스트 케이스 이름들이 명확하지 않은데 테스트 케이스만 봐도 어떤 요구사항을 구현했는지 알 수 있도록 만들어보세요.

briandr97 committed:
refactor(CarTest): 리뷰 반영 kotlin.runCatching -> runCatching

briandr97 committed:
refactor(CarTest): 리뷰 반영 테스트 케이스에 명확한 이름 부여

galcyurio reviewed (approved):
피드백 잘 반영해주셨네요 👍 
다음 단계도 힘내세요!
